### PR TITLE
perf: Replace blocking Mutex

### LINKED ISSUE #433: Replace PrecomputedCache's blocking Mutex with future Mutex
It's a subissue of #415

### PR DESCRIPTION
**Pull request by 0623forbidden**
Replace PrecomputedCache's blocking mutex with future Mutex
Fixes #433

### CODE REVIEW

File: tox_core/src/dht/server/mod.rs [hunk 1]
```diff
@@ -1208,87 +1254,113 @@ impl Server {
 
     /// Handle received `OnionRequest0` packet and send `OnionRequest1` packet
     /// to the next peer.
-    fn handle_onion_request_0(&self, packet: &OnionRequest0, addr: SocketAddr)
-        -> impl Future<Output = Result<(), HandlePacketError>> + Send {
+    fn handle_onion_request_0(&self, packet: OnionRequest0, addr: SocketAddr)
+                              -> impl Future<Output = Result<(), HandlePacketError>> + Send {
```
Comment (suhr) [inline tox_core/src/dht/server/mod.rs:1258]: I don't think changing the indentation here is necessary.

File: tox_core/src/net_crypto/mod.rs [hunk 2]
```diff
@@ -394,107 +393,135 @@ impl NetCrypto {
...
     /// Handle `CookieResponse` and if it's correct change connection status to `HandshakeSending`.
-    pub fn handle_cookie_response(&self, connection: &mut CryptoConnection, packet: &CookieResponse)
-        -> impl Future<Output = Result<(), HandlePacketError>> + Send {
-        let cookie_request_id = if let ConnectionStatus::CookieRequesting { cookie_request_id, .. } = connection.status {
-            cookie_request_id
-        } else {
-            return Either::Left(future::err(HandlePacketError::from(HandlePacketErrorKind::InvalidState)))
-        };
+    pub fn handle_cookie_response(&self, connection: Arc<RwLock<CryptoConnection>>, packet: CookieResponse)
+                                  -> impl Future<Output = Result<(), HandlePacketError>> + Send {
```
Comment (suhr) [inline tox_core/src/net_crypto/mod.rs:457]: `->` indentation.

**Review by suhr — CHANGES_REQUESTED** [review]
Overall, looks good to me.

**Review by suhr — APPROVED** [review]

File: tox_core/src/dht/precomputed_cache.rs [hunk 3]
```diff
@@ -28,15 +28,18 @@ impl PrecomputedCache {
...
 
     /// Get `PrecomputedKey` for the given `PublicKey`.
-    pub fn get(&self, pk: PublicKey) -> PrecomputedKey {
-        let mut keys = self.precomputed_keys.lock();
-
-        if let Some(precomputed_key) = keys.get(&pk) {
-            return precomputed_key.clone();
+    pub fn get(&self, pk: PublicKey) -> impl Future<Output = PrecomputedKey> + Send {
+        let precomputed_cache = self.clone();
+        async move {
```
Comment (kurnevsky) [inline tox_core/src/dht/precomputed_cache.rs:33]: Can it be replaced by `async fn`?

Comment (0623forbidden) [reply tox_core/src/dht/precomputed_cache.rs:33]: Shure.

File: tox_core/src/dht/server/mod.rs [hunk 4]
```diff
@@ -641,12 +641,16 @@ impl Server {
...
         };
-        let ping_req = Packet::PingRequest(PingRequest::new(
-            &self.precomputed_keys.get(node.pk),
-            &self.pk,
-            &payload
-        ));
-        self.send_to(node.saddr, ping_req)
+        let server = self.clone();
+        let node = *node;
+        async move {
```
Comment (kpp) [inline tox_core/src/dht/server/mod.rs:646]: this one can be async fn too

**Comment by kpp** [comment]
Look there are lots of places we can groom our code like with 'return move async {}` vs async fn`. Do you want to do it or shall we?

**Comment by 0623forbidden** [comment]
@kpp I want to do it (`async fn`) with #414 
Move to `async fn` brings to complicated errors (I tried already) even for simple cases as you mention in comment above. I think move to `async fn` is out of "Replace blocking mutex" issue.

**Comment by suhr** [comment]
Indeed, refactoring to `async fn` should be a separate PR.

**Comment by kurnevsky** [comment]
I'd vote for `async fn` in this pr because currently it introduces a lot of unnecessary clones, like https://github.com/tox-rs/tox/pull/434/files#diff-334f26e0ab7f6aa05e72ad3c625a85e7R703-R704

**Comment by kurnevsky** [comment]
Though just tried to do it myself and it seems it's indeed not trivial :)

File: tox_core/src/net_crypto/mod.rs [hunk 5]
```diff
@@ -1021,21 +1047,27 @@ impl NetCrypto {
 
     /// Send `CookieRequest` or `CryptoHandshake` packet if needed depending on
     /// connection status and update sent counter
-    fn send_status_packet(&self, connection: &mut CryptoConnection) -> impl Future<Output = Result<(), SendPacketError>> + Send {
-        match connection.packet_to_send() {
+    fn send_status_packet(&self, connection: Arc<RwLock<CryptoConnection>>, status: Option<ConnectionStatus>)
```
Comment (kurnevsky) [inline tox_core/src/net_crypto/mod.rs:1050]: Why did you change it everywhere to Arc<RwLock<..>>?

File: tox_core/src/net_crypto/mod.rs [hunk 6]
```diff
@@ -1021,21 +1047,27 @@ impl NetCrypto {
...
         }
     }
 
     /// Send `CryptoData` packet if the connection is established.
-    fn send_data_packet(&self, connection: &mut CryptoConnection, data: Vec<u8>, packet_number: u32)
+    fn send_data_packet(&self, connection: Arc<RwLock<CryptoConnection>>, data: Vec<u8>, packet_number: u32)
         -> impl Future<Output = Result<(), SendDataError>> + Send {
-        let packet = match connection.status {
+        let buffer_start = connection.read().recv_array.buffer_start;
+        let packet = match connection.write().status {
```
Comment (kurnevsky) [inline tox_core/src/net_crypto/mod.rs:1066]: I don't really like how it looks. First we acquire read lock, then release it and acquire write lock... Need to check whether we can deadlock this way.

File: tox_core/src/net_crypto/mod.rs [hunk 7]
```diff
@@ -1086,57 +1118,56 @@ impl NetCrypto {
...
-                    let status_future = self.send_connection_status(&connection, false)
+                if connection.read().is_established() {
+                    let status_future = self.send_connection_status(&connection.read(), false)
                         .map_err(|e| e.context(SendDataErrorKind::SendToConnectionStatus).into());
                     futures.push(Box::pin(status_future));
                 }
 
-                if connection.is_established() || connection.is_not_confirmed() {
-                    futures.push(Box::pin(self.send_kill_packet(&mut connection)));
+                if connection.read().is_established() || connection.read().is_not_confirmed() {
```
Comment (kurnevsky) [inline tox_core/src/net_crypto/mod.rs:1136]: A lot of lock acquires and releases...

File: tox_core/src/net_crypto/mod.rs (hunk 5 above)
Comment (0623forbidden) [reply tox_core/src/net_crypto/mod.rs:1050]: I can explain it like:

1. `NetCrypto.handle_cookie_response()` is need to take ownership of Connection in `async {}` block, 
    can't do this by `&mut` because it pass connection to `send_status_packet()` across `await` point.
    That's, `send_status_packet()` should to mutate connection, but can't take &mut.
    My first try was an Arc<Mutex<, but.. 
2. `NetCrypto.main_loop()` is a "source" of `Connection`'s for another NetCrypto methods like `send_status_packet()`
3. `NetCrypto.connections` has type of `Arc<RwLock<HashMap<PublicKey, Arc<RwLock<CryptoConnection>>>>>`, 
    that means that `Arc<RwLock<CryptoConnection>>` is natural type for Connection in NetCrypto's architecture

File: tox_core/src/net_crypto/mod.rs (hunk 6 above)
Comment (0623forbidden) [reply tox_core/src/net_crypto/mod.rs:1066]: Firs I got totally deadlock but then I inspect an all deals with lock acqure line-by-line and got succes. But yes, it's not seems good for me anyway: too much places wants to modify Connection state etc

File: tox_core/src/net_crypto/mod.rs (hunk 7 above)
Comment (0623forbidden) [reply tox_core/src/net_crypto/mod.rs:1136]: Seems like architectural-depend issue

**Comment by 0623forbidden** [comment]
> I'd vote for `async fn` in this pr because currently it introduces a lot of unnecessary clones, like https://github.com/tox-rs/tox/pull/434/files#diff-334f26e0ab7f6aa05e72ad3c625a85e7R733

# [derive (Clone)] brings to feel free to go easy way to deal with it :)

File: tox_core/src/net_crypto/mod.rs (hunk 5 above)
Comment (kurnevsky) [reply tox_core/src/net_crypto/mod.rs:1050]: Another reason to migrate to `async fn` in this MR - this will allow to leave it as `&mut`.

Comment (kurnevsky) [reply tox_core/src/net_crypto/mod.rs:1050]: You can check https://github.com/tox-rs/tox/commit/0dd558fd1a6766a153311f80453a4f73cb07158e where I partly migrated to `async fn` - I haven't figured out how yet to make echo example work, but it should be possible :)

**Comment by 0623forbidden** [comment]
@kurnevsky Well, I'll continue working on this PR for migrate to `async fn`

**Comment by kurnevsky** [comment]
Updated a bit `async_fn` branch - `echo` example should work after migrating `net_crypto` to `async fn`.

**Comment by kpp** [comment]
@kurnevsky you can push right into 0623forbidden:433_replace_blocking_mutex. Or do you prefer @0623forbidden to merge your changes into their PR?

**Comment by kurnevsky** [comment]
I can, if @0623forbidden doesn't mind.

**Comment by 0623forbidden** [comment]
@kurnevsky feel free to push right into 0623forbidden:433_replace_blocking_mutex
I'll rebase my local work

**Comment by 0623forbidden** [comment]
But wait a minute..

**Comment by 0623forbidden** [comment]
Sorry I forget to publish my wip
Please, rebase along my changes

**Comment by kurnevsky** [comment]
Why didn't you just take my commit? It contains a lot more changes: https://github.com/tox-rs/tox/compare/79cc5ff598c134f150cd54b4a27f917bf85b7dfc%5E...721aff6d70225ac976f49e867c85e902d833b55b

**Comment by 0623forbidden** [comment]
Oh, sorry, its f***ng "Large diffs are not rendered by default." 
Nevermind

**Comment by kurnevsky** [comment]
ok, so I pushed my commit here.

File: tox_core/src/dht/server_ext.rs [hunk 8]
```diff
@@ -26,7 +26,7 @@ impl ServerExt for Server {
         socket: UdpSocket,
         mut rx: Receiver<(Packet, SocketAddr)>,
         stats: Stats
-    ) -> Pin<Box<dyn Future<Output = Result<(), Error>> + Send>> {
+    ) -> Pin<Box<dyn Future<Output = Result<(), Error>>>> {
```
Comment (kpp) [inline tox_core/src/dht/server_ext.rs:29]: Can it be async fn?

File: examples/echo.rs [hunk 9]
```diff
@@ -58,190 +61,171 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+
+    futures::select!(
+        res = dht_server.run_socket(socket, rx, stats).fuse() => res.map_err(Error::from)?,
+        res = lan_discovery_sender.run().fuse() => res.map_err(Error::from)?,
+        res = tcp_connections.run().fuse() => res.map_err(Error::from)?,
+        res = onion_client.run().fuse() => res.map_err(Error::from)?,
+        res = net_crypto.run().fuse() => res.map_err(Error::from)?,
+        res = friend_connections.run().fuse() => res.map_err(Error::from)?,
+        res = net_crypto_tcp_future.fuse() => res?,
+        // res = tcp_incoming_future => res?,
```
Comment (kpp) [inline examples/echo.rs:224]: How so?

File: tox_core/src/dht/server_ext.rs (hunk 8 above)
Comment (kurnevsky) [reply tox_core/src/dht/server_ext.rs:29]: It can't - traits can't have async fns currently.

File: examples/echo.rs (hunk 9 above)
Comment (kurnevsky) [reply examples/echo.rs:224]: It doesn't compile now with this future, but should be fine once net_crypto is changed to use async fn.

Comment (kpp) [reply examples/echo.rs:224]: In the current PR or in another?

Comment (kurnevsky) [reply examples/echo.rs:224]: Current PR, otherwise it will break echo server :)

File: tox_core/src/dht/server_ext.rs (hunk 8 above)
Comment (0623forbidden) [reply tox_core/src/dht/server_ext.rs:29]: may be try to use async_trait crate?

**Comment by kurnevsky** [comment]
Pushed one more commit - now onion client is left + one test.

**Comment by kurnevsky** [comment]
ok, I think it can be merged in its current state. But need to be reviewed/tested first, as async fn changes semantic.

File: examples/dht_server.rs [hunk 10]
```diff
@@ -60,28 +61,18 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
     };
 
-    info!("Running DHT server on {}", local_addr);
+    if let Err(ref e) = res {
+        error!("Processing ended with error: {:?}", e)
+    }
 
-    let mut runtime = tokio::runtime::Runtime::new()?;
-    runtime.block_on(future)?;
+    info!("Running DHT server on {}", local_addr);
```
Comment (kpp) [inline examples/dht_server.rs:75]: This line should be somewhere upper

**Comment by kurnevsky** [comment]
I got rid of all aync move blocks, so now it looks good :)

File: examples/echo.rs [hunk 11]
```diff
@@ -58,190 +61,171 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
...
+    let tcp_incoming_future = tcp_incoming_future.map_err(|err| {
+        error!("Failed to handle packet: {:?}", err);
+        err
+    });
+
+    let net_crypto_c = net_crypto.clone();
+    let friend_connections_c = friend_connections.clone();
+    let lossless_future = async {
+        while let Some((pk, packet)) = lossless_rx.next().await {
+            let future = async {
```
Comment (kpp) [inline examples/echo.rs:160]: do we need this async block?

File: examples/dht_server.rs (hunk 10 above)
Comment (kurnevsky) [reply examples/dht_server.rs]: Right, fixed.

File: examples/echo.rs (hunk 11 above)
Comment (kurnevsky) [reply examples/echo.rs]: We don't, fixed.

File: tox_core/src/net_crypto/mod.rs [hunk 12]
```diff
@@ -1129,48 +1071,48 @@ impl NetCrypto {
...
-    pub fn run(self) -> impl Future<Output = Result<(), RunError>> + Send {
+    pub async fn run(&self) -> Result<(), RunError> {
         let mut wakeups = tokio::time::interval(PACKET_COUNTER_AVERAGE_INTERVAL);
 
-        async move {
-            while wakeups.next().await.is_some() {
-                let fut = tokio::time::timeout(
-                    PACKET_COUNTER_AVERAGE_INTERVAL, self.main_loop()
-                );
+        while let Some(_) = wakeups.next().await {
```
Comment (kpp) [inline tox_core/src/net_crypto/mod.rs:1097]:
```suggestion
        while wakeups.next().await.is_some() {
```

Comment (kurnevsky) [reply tox_core/src/net_crypto/mod.rs]: Fixed.

File: tox_core/src/relay/server/server.rs [hunk 13]
```diff
@@ -78,8 +78,6 @@ impl Server {
     handshaked client.
     */
     pub fn handle_packet(&self, pk: &PublicKey, packet: Packet) -> impl Future<Output = Result<(), Error>> + Send {
```
Comment (kpp) [inline tox_core/src/relay/server/server.rs:80]: How about make it async?

**Review by suhr — APPROVED** [review]
The code is much nicer when everything uses async.
